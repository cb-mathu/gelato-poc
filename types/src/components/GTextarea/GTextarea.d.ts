declare const _default: import("vue").DefineComponent<{
    label: {
        type: StringConstructor;
    };
    description: {
        type: StringConstructor;
    };
    onUpdate: import("vue").PropType<(event: Event) => void>;
    noResize: {
        type: BooleanConstructor;
        default: boolean;
    };
    id: {
        type: StringConstructor;
        required: boolean;
    };
    class: {
        type: StringConstructor;
        required: boolean;
        default: string;
    };
    disabled: {
        type: BooleanConstructor;
    };
    autofocus: {
        type: BooleanConstructor;
        default: boolean;
    };
    autocomplete: {
        type: StringConstructor;
        default: string;
    };
    required: {
        type: BooleanConstructor;
        default: boolean;
    };
    readonly: {
        type: BooleanConstructor;
        default: boolean;
    };
    name: {
        type: StringConstructor;
        default: string;
    };
    modelValue: {
        type: (StringConstructor | BooleanConstructor | NumberConstructor)[];
        default: string;
    };
    placeholder: {
        type: StringConstructor;
        default: string;
    };
    modelModifiers: {
        default: () => {};
    };
    invalid: {
        type: BooleanConstructor;
    };
    errorMessage: {
        type: StringConstructor;
    };
}, () => JSX.Element, unknown, {}, {}, import("vue").ComponentOptionsMixin, import("vue").ComponentOptionsMixin, Record<string, any>, string, import("vue").VNodeProps & import("vue").AllowedComponentProps & import("vue").ComponentCustomProps, Readonly<{
    class: string;
    autofocus: boolean;
    disabled: boolean;
    name: string;
    placeholder: string;
    modelValue: string | number | boolean;
    autocomplete: string;
    required: boolean;
    readonly: boolean;
    modelModifiers: {};
    invalid: boolean;
    noResize: boolean;
} & {
    id?: string | undefined;
    label?: string | undefined;
    errorMessage?: string | undefined;
    description?: string | undefined;
    onUpdate?: ((event: Event) => void) | undefined;
}>, {
    class: string;
    autofocus: boolean;
    disabled: boolean;
    name: string;
    placeholder: string;
    modelValue: string | number | boolean;
    autocomplete: string;
    required: boolean;
    readonly: boolean;
    modelModifiers: {};
    invalid: boolean;
    noResize: boolean;
}>;
export default _default;
