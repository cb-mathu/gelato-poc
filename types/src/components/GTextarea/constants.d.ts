import { PropType } from 'vue';
export declare const TEXTAREA_PROPS: {
    label: {
        type: StringConstructor;
    };
    description: {
        type: StringConstructor;
    };
    onUpdate: PropType<(event: Event) => void>;
    noResize: {
        type: BooleanConstructor;
        default: boolean;
    };
    id: {
        type: StringConstructor;
        required: boolean;
    };
    class: {
        type: StringConstructor;
        required: boolean;
        default: string;
    };
    disabled: {
        type: BooleanConstructor;
    };
    autofocus: {
        type: BooleanConstructor;
        default: boolean;
    };
    autocomplete: {
        type: StringConstructor;
        default: string;
    };
    required: {
        type: BooleanConstructor;
        default: boolean;
    };
    readonly: {
        type: BooleanConstructor;
        default: boolean;
    };
    name: {
        type: StringConstructor;
        default: string;
    };
    modelValue: {
        type: (StringConstructor | BooleanConstructor | NumberConstructor)[];
        default: string;
    };
    placeholder: {
        type: StringConstructor;
        default: string;
    };
    modelModifiers: {
        default: () => {};
    };
    invalid: {
        type: BooleanConstructor;
    };
    errorMessage: {
        type: StringConstructor;
    };
};
