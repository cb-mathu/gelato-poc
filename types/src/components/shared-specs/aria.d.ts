export declare const FormFieldAria: (Component: any, label: string, role: string) => void;
export declare const ControlFieldAria: (Component: any, label: string, role: string) => void;
export declare const InputFieldAria: (Component: any, label: string, role: string) => void;
