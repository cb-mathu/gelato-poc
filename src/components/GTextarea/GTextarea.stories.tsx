import { ref } from 'vue';
import GTextarea from './GTextarea';
import { GNotification } from '../GNotification';
import { MergedTextareaProps } from '../core/Textarea/types';

export default {
  title: 'Component/Forms/GTextarea',
  component: GTextarea,
  argTypes: {
    name: {
      default: 'Name of the field',
    },
    label: {
      description: 'Label for the field',
      control: {
        type: 'text',
      },
    },
    description: {
      description: 'Description for the field',
      control: {
        type: 'text',
      },
    },
    disabled: {
      description: 'Disable the field',
      control: {
        type: 'boolean',
      },
    },
    autofocus: {
      description: 'Autofocus the field',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    autocomplete: {
      description: 'Autcomplete for the field',
      defaultValue: 'off',
      control: {
        type: 'text',
      },
    },
    required: {
      description: 'Field is required',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    readonly: {
      description: 'Field is readonly',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    modelValue: {
      description: 'v-model value',
      control: {
        type: 'text',
      },
    },
    id: {
      description: 'Id of the field',
      control: {
        type: 'text',
      },
    },
    placeholder: {
      description: 'placeholder of the field',
      control: {
        type: 'text',
      },
    },
    class: {
      description: 'Class of the field',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
    invalid: {
      description: 'Field is invalid',
      defaulValue: false,
      control: {
        type: 'boolean',
      },
    },
    errorMessage: {
      description: 'Error message for invalid field (Applicable only if invalid is set)',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
    maxlength: {
      description: 'Min value (Applicable for Text field)',
      control: {
        type: 'number',
      },
    },
    noResize: {
      description: 'Do not Allow resize for the field',
      defaultVal: false,
      control: {
        type: 'boolean',
      },
    },
    cols: {
      description: 'number of cols for the field',
      control: {
        type: 'number',
      },
    },
  },
};

const Template = (args: any) => ({
  setup() {
    let input = ref('Hello from chargebee!');

    const attributes = {
      disabled: args.disabled,
      readonly: args.readonly,
      invalid: args.invalid,
      errorMessage: args.errorMessage,
    };

    return () => (
      <div>
        <div style="margin-bottom: 20px;">
          <GTextarea {...attributes} v-model={input.value} id="input" label={args.label} description={args.description}>
            {{
              label: () => args.labelSlot,
              default: () => args.descriptionSlot,
            }}
          </GTextarea>
        </div>
        <GNotification>
          Entered Value:<b>{input.value}</b>
        </GNotification>
      </div>
    );
  },
});

//Default
export const Story = Template.bind({});


const testLabel = 'First Name';
const testDescription  = 'Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin';

export const DefaultTextarea = (props?: MergedTextareaProps, slots?: any) => (
  <GTextarea key="default" modelValue='Hello from chargebee' {...(props as any)} v-slots={slots} />
);


// With Label
export const TextareaWithLabel = (props?:any) => <GTextarea modelValue='Hello from chargebee' label={testLabel} {...props}/>

// With description
export const TextareaWithDescription = () => <GTextarea label={testLabel} modelValue='Hello from chargebee' description={testDescription}/>

// with Slots
export const TextareaWithSlots: any = (props?:any) => <GTextarea modelValue='Hello from chargebee' {...props} v-slots={{label: ()=> testLabel, default: () => testDescription }} />

// Disabled
export const DisabledTextarea: any = ()=> <GTextarea modelValue='Hello from chargebee' label={testLabel} disabled/>

// Readonly
export const ReadonlyTextarea: any = () =>  <GTextarea modelValue='Hello from chargebee' label={testLabel} readonly/>

// Invalid
export const InvalidTextarea: any = () =>  <GTextarea modelValue='Hello from chargebee' label={testLabel} invalid errorMessage="This is an error message, ooh poor you!"/>

// Lazy
export const LazyTextarea = () => ({
  setup() {
    let input = ref('Hello from chargebee');

    return () => (
      <div>
        <div style="margin-bottom: 20px;">
          <GTextarea label={testLabel} v-model={[input.value, ['lazy']]} />
        </div>
        <GNotification>
          Entered Value:<b>{input.value}</b>
        </GNotification>
      </div>
    );
  },
});