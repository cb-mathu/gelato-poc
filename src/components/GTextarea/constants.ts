import { PropType } from 'vue';
import { CORE_TEXTAREA_PROPS } from '../core/Textarea';

export const TEXTAREA_PROPS = {
  ...CORE_TEXTAREA_PROPS,
  label: {
    type: String,
  },
  description: {
    type: String,
  },
  onUpdate: Function as PropType<(event: Event) => void>,
};
