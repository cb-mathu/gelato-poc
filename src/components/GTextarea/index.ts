import GTextarea from './GTextarea';

export { GTextarea };
