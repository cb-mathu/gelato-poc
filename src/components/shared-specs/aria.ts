/* istanbul ignore file */
import { render } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import { screen } from '@testing-library/dom';

export const FormFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Form Field ARIA`, () => {
    test('renders aria-labelledBy', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveAttribute('aria-labelledBy');
    });

    test('renders aria-invalid', async () => {
      render(Component({ invalid: true }));
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toBeInvalid();
    });

    test('renders aria-errorMessage', async () => {
      const errorMessage = 'This is an error message, ooh poor you!';
      render(Component({ invalid: true, errorMessage: errorMessage }));
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveErrorMessage(errorMessage);
    });
  });

export const ControlFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Control Field ARIA`, () => {
    test('renders aria-checked', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByRole(role, { checked: true })).toBeTruthy();
    });
    FormFieldAria(Component, label, role);
  });

export const InputFieldAria = (Component: any, label: string, role: string) =>
  describe(`G${role.toUpperCase()}-Input Field ARIA`, () => {
    test('renders aria-describedBy', async () => {
      render(Component());
      await userEvent.click(screen.getByLabelText(label));
      expect(screen.getByLabelText(label)).toHaveAccessibleDescription();
    });
    FormFieldAria(Component, label, role);
  });
