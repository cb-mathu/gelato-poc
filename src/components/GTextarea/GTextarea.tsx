import { defineComponent } from 'vue';
import { Textarea } from '../core/Textarea';
import { TEXTAREA_PROPS } from './constants';
import { useFormAria, useFormAriaIds } from '../../utils/accessibility';
import { GInputErrorMessage } from '../GInput';
import { isNotEmpty } from '../../utils/common';

export default defineComponent({
  name: 'GTextarea',
  props: TEXTAREA_PROPS,

  setup(props, { attrs, slots }) {
    const { labelId, descriptionId, errorId } = useFormAriaIds(props.id);

    return () => {
      const attributes = {
        ...attrs,
        ...props,
      };

      const classes = {
        [`g-input`]: true,
        [`g-input--invalid`]: props.invalid,
      };

      const label = props.label || slots.label?.();
      const description = props.description || slots.default?.();

      const ariaAttributes = {
        ...useFormAria({
          labelledBy: labelId,
          describedBy: descriptionId,
          invalid: props.invalid,
          errorMessage: errorId,
        }),
      };

      return (
        <div class={classes}>
          {isNotEmpty(label) && (
            <label id={labelId} for={attributes.id} class="g-input__label">
              {label}
            </label>
          )}
          {isNotEmpty(description) && (
            <div id={descriptionId} class="g-input__desc">
              {description}
            </div>
          )}
          <div class="g-input__group">
            <Textarea {...attributes} {...ariaAttributes} class="g-input__element" />
          </div>
          {props.invalid && props.errorMessage && <GInputErrorMessage id={errorId} errorMessage={props.errorMessage} />}
        </div>
      );
    };
  },
});
