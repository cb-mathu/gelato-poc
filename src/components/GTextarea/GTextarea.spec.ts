import { render } from '@testing-library/vue';
import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';
import * as stories from './GTextarea.stories';
import { InputFieldAria } from '../shared-specs/aria';
const {
  DefaultTextarea,
  TextareaWithLabel,
  TextareaWithDescription,
  TextareaWithSlots,
  DisabledTextarea,
  ReadonlyTextarea,
  InvalidTextarea,
  LazyTextarea,
} = stories;

const testLabel = 'First Name';
const testErroMessage = 'This is an error message, ooh poor you!';

describe('GTextarea: ', () => {
  test('renders with default args', () => {
    const { container } = render(DefaultTextarea());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders description', () => {
    const { container } = render(TextareaWithDescription());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders slots', () => {
    const { container } = render(TextareaWithSlots());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders in disabled state', () => {
    render(DisabledTextarea());
    expect(screen.getByLabelText(testLabel)).toHaveAttribute('disabled');
  });

  test('renders in readonly state', () => {
    render(ReadonlyTextarea());
    expect(screen.getByLabelText(testLabel)).toHaveAttribute('readonly');
  });

  test('renders in invalid state', () => {
    render(InvalidTextarea());
    expect(screen.getByText(testErroMessage)).toBeVisible();
  });

  test('renders cols', () => {
    render(
      TextareaWithLabel({
        cols: 5,
      }),
    );
    expect(screen.getByLabelText(testLabel)).toHaveAttribute('cols');
  });

  test('renders rows', () => {
    render(
      TextareaWithLabel({
        rows: 5,
      }),
    );
    expect(screen.getByLabelText(testLabel)).toHaveAttribute('rows');
  });

  test('pressing tab will focus textarea', () => {
    render(TextareaWithLabel());
    userEvent.tab();
    expect(screen.getByLabelText(testLabel)).toHaveFocus();
  });

  test('focuses textarea when label is clicked', () => {
    render(TextareaWithLabel());
    const element = screen.getByText(testLabel);
    userEvent.click(element);
    expect(screen.getByLabelText(testLabel)).not.toHaveFocus();
  });

  test('lazy: updates textarea on change event', () => {
    const { container } = render(LazyTextarea());
    const inputValue = ' to user';
    userEvent.type(screen.getByLabelText(testLabel), inputValue);
    container.parentElement?.click();
    expect(screen.getByLabelText(testLabel)).toHaveValue('Hello from chargebee to user');
  });
});

InputFieldAria(TextareaWithSlots, testLabel, 'textarea');
