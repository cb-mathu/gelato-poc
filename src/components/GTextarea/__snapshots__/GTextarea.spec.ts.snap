// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GTextarea:  renders description 1`] = `
<div
  class="g-input"
>
  <label
    class="g-input__label"
    id="label_5"
  >
    First Name
  </label>
  <div
    class="g-input__desc"
    id="description_6"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin
  </div>
  <div
    class="g-input__group"
  >
    <textarea
      aria-describedby="description_6"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_5"
      autocomplete="off"
      class="g-input__element"
      description="Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin"
      id="CoreTextarea_8"
      invalid="false"
      label="First Name"
      modelmodifiers="[object Object]"
      modelvalue="Hello from chargebee"
      name=""
      noresize="false"
      placeholder=""
    />
  </div>
  <!---->
</div>
`;

exports[`GTextarea:  renders slots 1`] = `
<div
  class="g-input"
>
  <label
    class="g-input__label"
    id="label_9"
  >
    
    First Name
    
  </label>
  <div
    class="g-input__desc"
    id="description_10"
  >
    
    Lorem ipsum dolor sit amet, consectetur adipiscing elitroin.Lorem ipsum dolor sit amet, consectetur adipiscing elitroin
    
  </div>
  <div
    class="g-input__group"
  >
    <textarea
      aria-describedby="description_10"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_9"
      autocomplete="off"
      class="g-input__element"
      id="CoreTextarea_12"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="Hello from chargebee"
      name=""
      noresize="false"
      placeholder=""
    />
  </div>
  <!---->
</div>
`;

exports[`GTextarea:  renders with default args 1`] = `
<div
  class="g-input"
>
  <!---->
  <!---->
  <div
    class="g-input__group"
  >
    <textarea
      aria-describedby="description_2"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_1"
      autocomplete="off"
      class="g-input__element"
      id="CoreTextarea_4"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue="Hello from chargebee"
      name=""
      noresize="false"
      placeholder=""
    />
  </div>
  <!---->
</div>
`;
